onnxruntime/contrib_ops/cuda/math/bias_softmax_impl.cu

```diff
@@ -0,0 +1,314 @@
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/math/softmax_impl.cuh"
+#include "core/providers/cuda/cu_inc/binary_elementwise_impl.cuh"
+#include "core/providers/cuda/math/binary_elementwise_ops_impl_functors.cuh"
+#include "bias_softmax.h"
+
+#include <limits>
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+// Duplicated softmax_impl.cu here
+// So far attempt to use shared kernel with additional template resulted in lost performance
+
+// Note: The intended case for 'input_bias' is the input sequence mask for transformer models
+// As an additive mask, it should be zero for preserved tokens and -infty for tokens to screen
+// The mask will broadcast from [batch_size, 1, 1, seq_len] to input [batch_size, num_heads, seq_len, seq_len]
+// Here element_count = seq_len and bias_broadcast_size_per_batch = num_heads * seq_len
+
+// The softmax + additive mask fusion follows NVIDIA apex's additive_masked_softmax_warp_forward
+// see https://github.com/NVIDIA/apex/blob/master/apex/contrib/csrc/multihead_attn/softmax.h
+
+template <typename input_t, typename output_t, typename acc_t, int log2_elements>
+__global__ void bias_softmax_warp_forward(
+  output_t* output, 
+  const input_t* input, 
+  const input_t* input_bias,
+  int element_count,
+  int batch_count, 
+  int batch_stride,  
+  int bias_broadcast_count_per_batch) {
+
+  // "WARP" refers to cooperative threads and might not equal 32 threads of GPU warp
+  // thread block is (WARP_SIZE, 128/WARP_SIZE)
+  constexpr int next_power_of_two = 1 << log2_elements;
+  constexpr int WARP_SIZE = (next_power_of_two < GPU_WARP_SIZE) ? next_power_of_two : GPU_WARP_SIZE;
+  constexpr int WARP_ITERATIONS = next_power_of_two / WARP_SIZE;
+  constexpr int WARP_BATCH = (next_power_of_two <= 128) ? 2 : 1;
+
+  // each "WARP" (<=32) processes WARP_BATCH(one of {1,2}) batches
+  int first_batch = (blockDim.y * blockIdx.x + threadIdx.y) * WARP_BATCH;
+
+  // last warp may have fewer batches
+  int local_batches = batch_count - first_batch;
+  if (local_batches > WARP_BATCH)
+    local_batches = WARP_BATCH;
+
+  // thread will process elements (local_index + n * warp_size) within batch
+  int local_idx = threadIdx.x;
+
+  // push input, input_bias output pointers to batch we need to process
+  input  += first_batch * batch_stride + local_idx;
+  output += first_batch * batch_stride + local_idx;
+
+  // load from global memory and apply bias (likely an additive mask)
+  acc_t elements[WARP_BATCH][WARP_ITERATIONS];
+  for (int i = 0; i < WARP_BATCH; ++i) {
+
+    // the bias has assumed shape [batch_size, element_count] 
+    // .. and needs to broadcast to [batch_size, broadcast_size, element_count]
+    int bias_offset = (first_batch + i)/bias_broadcast_count_per_batch * batch_stride + local_idx; 
+
+    int batch_element_count = (i >= local_batches) ? 0 : element_count;
+    for (int it = 0; it < WARP_ITERATIONS; ++it) {
```

Comment:
>WARP_ITERATIONS [](start = 26, length = 15)

Please, comment on what other values of WARP_ITERATIONS you considered and which one gives better perf.

Reply:
@yuslepukhin The body of this routine was taken from softmax_impl.cu. That in turn was taken from pytorch's implementation of softmax. I did not evaluate the performance w.r.t. to free parameters.

Reply:
Going to close this as it can be considered a separate work item re: refactoring and maximizing softmax performance itself (not just biased version).